odoo/tools/urls.py

```diff
@@ -0,0 +1,75 @@
+import re
+import urllib.parse
+from urllib.parse import _WHATWG_C0_CONTROL_OR_SPACE
+
+__all__ = ['urljoin']
+
+
+def _contains_dot_segments(path: str) -> str:
+    # most servers decode url before doing dot segment resolutions
+    decoded_path = urllib.parse.unquote(path, errors='strict')
+    return any(seg in ('.', '..') for seg in decoded_path.split('/'))
+
+
+def urljoin(base: str, extra: str) -> str:
+    """Join a trusted base URL with a relative URL safely.
+
+    Unlike standard URL joins that follow RFC 3986 (e.g., `urllib.parse.urljoin`),
+    this function enforces strict behavior that better aligns with developer
+    expectations and guards against path traversals, unplanned redirects, and
+    accidental host/scheme overrides.
+
+    - Behaves similarly to `base + '/' + extra`
+    - Keeps scheme and netloc from `base`, and raises an error if `extra` has them
+    - Ignores any scheme/host in `extra`
+    - Forbids `.` and `..` path traversal
+    - merges path/query/fragment
+
+    :param base: Trusted base URL or path.
+    :type base: str
+    :param extra: Relative URL (`path`, `?query`, `#frag`). No scheme & host allowed unless it matches `base`
+    :type extra: str
+    :returns: joined URL.
+    :rtype: str
```

Comment:
Type information is already in signature.
```suggestion
    :param base: Trusted base URL or path.
    :param extra: Relative URL (`path`, `?query`, `#frag`). No scheme & host allowed unless it matches `base`
    :returns: joined URL.
```